### PR TITLE
Agent mode: frontend

### PR DESCRIPTION
**Pull request by ngafar**
# Description

This PR includes:

- Added a dropdown to toggle between agent mode and chat.
- Agent messages are now formatted.
- Agent messages can be edited. 
- You can now execute agent's plan.

What's not covered:

- 1st message always has a code block that says "undefined."

# Testing

Clone and test.

# Documentation

Not yet, waiting for final release.

### CODE REVIEW

File: mito-ai/src/components/DropdownMenu.tsx [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+            {isOpen && (
+                <div className="dropdown-menu">
+                    {items.map((item, index) => (
+                        <button
+                            key={index}
+                            className={`dropdown-item ${item.icon ? 'dropdown-item-with-icon' : ''}`}
+                            onClick={() => handleItemClick(item.onClick)}
+                            style={{ display: 'flex', gap: '5px', alignItems: 'center' }}
+                        >
+                            {item.icon && (
```
Comment (aarondr77) [inline mito-ai/src/components/DropdownMenu.tsx:52]: Let's make sure that the dropdown item icon container always takes up a consistent width. Like: 20px or whatever regardless of the size of the icon inside. 

Right now the text is not aligned. 
Screenshot 2025-02-07 at 1 06 05 PM

File: mito-ai/src/icons/RobotHeadIcon.tsx [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+    fill = '#E9E0FD',
+    height = '10',
+    width = '16',
+}) => (
+        <svg
+            data-testid="robot-head-icon"
+            xmlns="http://www.w3.org/2000/svg"
+            width={width}
+            height={height}
+            viewBox="0 0 16 10"
```
Comment (aarondr77) [inline mito-ai/src/icons/RobotHeadIcon.tsx:17]: Same update here

File: mito-ai/src/icons/RobotHeadIcon.tsx [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+    height?: string;
+    width?: string;
+}> = ({
+    fill = '#E9E0FD',
+    height = '10',
+    width = '16',
+}) => (
+        <svg
+            data-testid="robot-head-icon"
+            xmlns="http://www.w3.org/2000/svg"
```
Comment (aarondr77) [inline mito-ai/src/icons/RobotHeadIcon.tsx:14]: Do we need these?

File: mito-ai/src/icons/RobotHeadIcon.tsx [hunk 4]
```diff
@@ -0,0 +1,30 @@
+import React from 'react';
+
+const RobotHeadIcon: React.FC<{
+    fill?: string;
```
Comment (aarondr77) [inline mito-ai/src/icons/RobotHeadIcon.tsx:4]: Let's do the same updating of fill through the css instead of passing the color

File: mito-ai/style/ChatTaskpane.css [hunk 5]
```diff
@@ -57,13 +58,20 @@
...
     color: var(--chat-assistant-message-font-color);
 }
 
+.message-agent {
+    background-color: var(--purple-300);
+}
+
+.message-agent .jp-RenderedHTMLCommon {
+    color: var(--grey-900) !important;
+}
```
Comment (aarondr77) [inline mito-ai/style/ChatTaskpane.css:73]: What is the `.jp-RenderedHTMLCommon` here? 

1. I would have thought that we wanted to use the same color for both of these `message-agents`, but I guess the .jp-RenderedHTMLCommon is some modifier?
2. I think using --purple-300 is fine since that's our custom color, but for the greys, lets try to use the jupyter colors like .jp-input whatever so that it is reactive to the theme changes.

File: mito-ai/style/ChatTaskpane.css [hunk 6]
```diff
@@ -126,6 +134,7 @@
 .message-edit-button {
     background-color: rgba(255, 255, 255, 0);
     border: none;
+    color: var(--grey-900) !important;
```
Comment (aarondr77) [inline mito-ai/style/ChatTaskpane.css:137]: Same for using the jupyter colors here

File: mito-ai/style/DropdownMenu.css [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+
+.dropdown-menu {
+    position: absolute;
+    top: calc(100% + 4px);
+    left: 0;
+    z-index: 1000;
+    background-color: var(--jp-layout-color1);
+    border: 1px solid var(--jp-layout-color2);
+    border-radius: 3px;
+    box-shadow: 0 2px 4px rgba(0, 0, 0, 0.2);
```
Comment (aarondr77) [inline mito-ai/style/DropdownMenu.css:14]: Nice touch! The shadow looks good!

File: mito-ai/src/icons/RobotHeadIcon.tsx [hunk 8]
```diff
@@ -0,0 +1,30 @@
+import React from 'react';
+
+const RobotHeadIcon: React.FC<{
```
Comment (aarondr77) [inline mito-ai/src/icons/RobotHeadIcon.tsx:3]: I think in practice the robot icon might be a little too like detailed for the size we are displaying it at? 
Screenshot 2025-02-07 at 1 06 05 PM

Maybe we can find a slightly less detailed icon and one that is more symmetrical in height and width. it looks a bit strange since they take up different amounts of space naturally

File: mito-ai/src/components/DropdownMenu.tsx [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+                {trigger}
+            </div>
+            {isOpen && (
+                <div className="dropdown-menu">
+                    {items.map((item, index) => (
+                        <button
+                            key={index}
+                            className={`dropdown-item ${item.icon ? 'dropdown-item-with-icon' : ''}`}
+                            onClick={() => handleItemClick(item.onClick)}
+                            style={{ display: 'flex', gap: '5px', alignItems: 'center' }}
```
Comment (aarondr77) [inline mito-ai/src/components/DropdownMenu.tsx:50]: These actually are more aligned if we remove the `alignItems` param

Currently 
Screenshot 2025-02-07 at 1 06 05 PM

vs this without it 

Screenshot 2025-02-07 at 1 15 46 PM

File: mito-ai/src/Extensions/AiChat/ChatTaskpane.tsx [hunk 10]
```diff
@@ -277,6 +300,58 @@ const ChatTaskpane: React.FC<IChatTaskpaneProps> = ({
...
+                `Install the following dependencies: ${agentResponse.dependencies.join(', ')}`,
+                'agent:planning'
+            )
+        }
+
+        // Loop through each action in the agent response 
+        // and add it to the chat history.
+        agentResponse.actions.forEach((action: string) => {
+            newChatHistoryManager.addAIMessageFromResponse(
+                action,
```
Comment (aarondr77) [inline mito-ai/src/Extensions/AiChat/ChatTaskpane.tsx:318]: As we are making these, let's add "Step 1: ", "Step 2:"  etc

File: mito-ai/src/Extensions/AiChat/ChatHistoryManager.tsx [hunk 11]
```diff
@@ -98,7 +98,7 @@ export class ChatHistoryManager {
         }
     }
 
-    updateMessageAtIndex(index: number, newContent: string): IOutgoingMessage {
+    updateMessageAtIndex(index: number, newContent: string, isAgentMessage: boolean = false): IOutgoingMessage {
```
Comment (aarondr77) [inline mito-ai/src/Extensions/AiChat/ChatHistoryManager.tsx:101]: If I update the original message I send to the agent, I want it to rerun the agent:planning stuff. But currently, it seems to just send it as a regular chat message

File: mito-ai/src/Extensions/AiChat/ChatTaskpane.tsx [hunk 12]
```diff
@@ -81,8 +86,7 @@ const ChatTaskpane: React.FC<IChatTaskpaneProps> = ({
     // Add this ref for the chat messages container
     const chatMessagesRef = useRef<HTMLDivElement>(null);
 
-    // TODO: RE-ADD setAgentModeEnabled when we need to call it. 
-    const [agentModeEnabled, _] = useState<boolean>(false)
+    const [agentModeEnabled, setAgentModeEnabled] = useState<boolean>(false)
```
Comment (aarondr77) [inline mito-ai/src/Extensions/AiChat/ChatTaskpane.tsx:89]: Currently the agent messages are getting logged as `mito_ai_chat_success` I think what we want is: 

1. `mito_ai_agent_prompt_success` for the first message that the user prompts to the agent
2. `mito_ai_agent_execution_success` for the future messages that the agent sends to execute it's plan

**Review by aarondr77 — COMMENTED** [review]
👏🏻✨Woohooo great job this. It feels super sweet and definetly way more efficient then iterating with the AI myself. So far everything I've tested it with is EDA styled analyses and I think it's a great starting place for that. TBD on like "I need to recreate this exact report in this exact format" but we'll see! 

The functionality is working really well! This is a little bit of a bigger review than normal, but its mostly just small nitpicks that I think will help us maintain and continue developing this going forward, as well as a bunch of polish finishing touches that will be nice. 

Awesome work!

File: mito-ai/src/Extensions/AiChat/ChatTaskpane.tsx [hunk 13]
```diff
@@ -207,10 +211,19 @@ const ChatTaskpane: React.FC<IChatTaskpaneProps> = ({
...
         }, 100);
+
+        return true
     }
 
-    const handleUpdateMessage = async (messageIndex: number, newContent: string) => {
-        sendChatInputMessage(newContent, messageIndex)
+    const handleUpdateMessage = async (messageIndex: number, newContent: string, isAgentMessage: boolean = false) => {
+        if (isAgentMessage) {
+            // Update the agent message locally without sending it to the AI
```
[the remark this answers is not in the document]
Comment (ngafar) [reply mito-ai/src/Extensions/AiChat/ChatTaskpane.tsx:220]: Good call. Never actually though about comments, but this totally makes sense.

File: mito-ai/mito_ai/utils/open_ai_utils.py [hunk 14]
```diff
@@ -112,4 +112,9 @@ def get_open_ai_completion_function_params(
     if model == "gpt-4o-mini":
         completion_function_params["temperature"] = 0.0
 
+    # Remove "stream" key if response_format is set.
+    # beta.chat.completions.parse will error if we try to pass it as a param. 
+    if response_format:
```
[the remark this answers is not in the document]
Comment (ngafar) [reply mito-ai/mito_ai/utils/open_ai_utils.py:117]: I'm going to tackle this in a second PR.

File: mito-ai/src/icons/ChatIcon.tsx [hunk 15]
```diff
@@ -0,0 +1,25 @@
+import React from 'react';
+
+const ChatIcon: React.FC<{
+    fill?: string;
```
[the remark this answers is not in the document]
Comment (ngafar) [reply mito-ai/src/icons/ChatIcon.tsx:4]: Took me a sec to figure out what you were talking about here, but this is a good call. Small change, but they add up.

File: mito-ai/style/ChatTaskpane.css (hunk 5 above)
Comment (ngafar) [reply mito-ai/style/ChatTaskpane.css:78]: `.jp-RenderedHTMLCommon` appears in a div that sits between `message-agent` and the `p` tag that holds the actual message content. For simplicity sake, I've update this to be `.message-agent p`.

Comment (ngafar) [reply mito-ai/style/ChatTaskpane.css:78]: Using the jupyter colors here won't work because of the shade of purple we're using. In light mode things look fine, but in dark mode the white text is too light:

Screenshot 2025-02-10 at 10 25 44 AM

File: mito-ai/style/ChatTaskpane.css (hunk 6 above)
Comment (ngafar) [reply mito-ai/style/ChatTaskpane.css:142]: Screenshot 2025-02-10 at 10 25 44 AM

File: mito-ai/src/Extensions/AiChat/ChatTaskpane.tsx [hunk 16]
```diff
@@ -542,26 +617,58 @@ const ChatTaskpane: React.FC<IChatTaskpaneProps> = ({
...
+            label: 'Chat',
+            onClick: () => {
+                clearChatHistory()
+                setAgentModeEnabled(false);
+            },
+            icon: ChatIcon,
+            iconColor: 'var(--jp-ui-font-color1)'
+        },
+        {
+            label: 'Agent',
```
[the remark this answers is not in the document]
Comment (ngafar) [reply mito-ai/src/Extensions/AiChat/ChatTaskpane.tsx:637]: Added support for disable state to DropdownMenu.tsx. Going to tackle this in another PR.

File: mito-ai/src/Extensions/AiChat/ChatTaskpane.tsx (hunk 12 above)
Comment (ngafar) [reply mito-ai/src/Extensions/AiChat/ChatTaskpane.tsx:89]: Going to tackle this in a separate PR.

File: mito-ai/src/Extensions/AiChat/ChatTaskpane.tsx [hunk 17]
```diff
@@ -316,6 +349,62 @@ const ChatTaskpane: React.FC<IChatTaskpaneProps> = ({
         }
     }
 
+    const handleAgentResponse = (
```
Comment (aarondr77) [inline mito-ai/src/Extensions/AiChat/ChatTaskpane.tsx:352]: After building the plan, if I refresh my browser, the chat plan of attack is not dispalyed properly, I just get the json back. 

I think all we need to do is call the `handleAgentReponse` in the fetchHistory function when the message we get back is of type agent:planning. 

ie: we should update this code 

```typescript
// 4. Add messages to the ChatHistoryManager
history.forEach(item => {
    newChatHistoryManager.addChatMessageFromHistory(item);
});
```

Screenshot 2025-02-10 at 1 34 08 PM

Comment (aarondr77) [reply mito-ai/src/Extensions/AiChat/ChatTaskpane.tsx:352]: This won't add back the LetsGo button, but let's ignore that edge case for now. 

So: we want to render all of the planning messages. They won't be purple, but that's okay for now.

**Review by aarondr77 — APPROVED** [review]
After tests pass + you fix up this one comment LGTM
